### PR TITLE
Collapse sibling text nodes.

### PR DESCRIPTION
**Pull request by dmjio**
@niteria @FPtje @tysonzero 

This preserves the isomorphism between the DOM and HTML by normalizing both structures by collapsing sibling text nodes. It should* solve the "consecutive text node" problem during pre-rendering.

Solves #589 #520 .

### CODE REVIEW

File: ghc-src/Miso/Html/Internal.hs [hunk 1]
```diff
@@ -114,7 +114,19 @@ instance L.ToHtml (VTree action) where
...
+          = foldMap L.toHtml
+          . V.fromList
+          . collapseSiblingTextNodes
+          . V.toList
+          $ vChildren
+
+collapseSiblingTextNodes :: [VTree a] -> [VTree a]
+collapseSiblingTextNodes [] = []
+collapseSiblingTextNodes (VText x : VText y : xs) =
+  collapseSiblingTextNodes (VText (x <> " " <> y) : xs)
```
Comment (FPtje) [inline ghc-src/Miso/Html/Internal.hs:127]: Should there really be a space in between? I would expect multiple text nodes to be concatenated directly.

File: jsbits/util.js [hunk 2]
```diff
@@ -12,3 +12,16 @@ window['callBlur'] = function callBlur(id) {
     if (ele && ele.blur) ele.blur()
   }, 50);
 }
+
+window['collapseSiblingTextNodes'] = function collapseSiblingTextNodes(vs) {
+  if (!vs) { return []; }
+  var ax = 0, adjusted = vs.length > 0 ? [vs[0]] : [];
+  for (var ix = 1; ix < vs.length; ix++) {
+    if (adjusted[ax]['type'] === 'vtext' && vs[ix]['type'] === 'vtext') {
+	adjusted[ax]['text'] += ' ' + vs[ix]['text'];
```
Comment (FPtje) [inline jsbits/util.js:21]: Same here

File: frontend-src/Miso/Html/Internal.hs [hunk 3]
```diff
@@ -130,7 +130,10 @@ node ns tag key attrs kids = View $ \sink -> do
   set "tag" tag vnode
   set "key" key vnode
   setAttrs vnode sink
-  flip (set "children") vnode =<< ghcjsPure . jsval =<< setKids sink
+  flip (set "children") vnode
+    =<< collapseSiblingTextNodes
```
Comment (FPtje) [inline frontend-src/Miso/Html/Internal.hs:134]: It looks like the children are being collapsed in ghcjs before every diff. Is this necessary? Can we get away with doing it only on the initial sync up between the server sent html and first render?

I'm not sure there were diffing problems between double text nodes when not using isomorphism, so feel free to shoot this idea down immediately 😅

Comment (dmjio) [reply frontend-src/Miso/Html/Internal.hs]: This is a good idea, we shouldn't call collapse on each diff iteration.

File: jsbits/isomorphic.js [hunk 4]
```diff
@@ -1,19 +1,40 @@
...
   }
+  if (debug) {
+    console.info ('Successfully prendered page');
+  }
   return true;
 }
 
-window['walk'] = function walk(vtree, node, doc) {
+window['diagnoseError'] = function diagnoseError(debug, vtree, node) {
+    if (debug) console.warn('VTree differed from node', vtree, node);
```
Comment (FPtje) [inline jsbits/isomorphic.js:34]: I'm a bit skeptic about the `debug` variable here. The name of the variable (`debug`) does not immediately make clear *what* it debugs. Documentation has to be read.

Besides, what it does is show these debug prints. Is that something we actually expect end users to use? Is it easy enough to mismatch and hard enough to debug such mismatches? Biggest cause for mismatches AFAIK are those adjacent text nodes. That's exactly what you fix.

Perhaps we should look at a different means of debugging such mismatches, maybe some wrapper around the `miso` function?

Comment (niteria) [reply jsbits/isomorphic.js:34]: I've used `miso` for a couple of weeks and have run into multiple different causes for the mismatches, including:

* The problem you mention - adjacent text nodes
* Different Route URIs: #587
* Bug in the framework that renders `br_` differently: #585
* My own mistakes where I didn't ensure that the Model on the server is the same that the model I use in the client

Each time I had to resort to commenting out pieces of the code to narrow it down and before the fallback code was fixed in this pull request the app would just hang with no feedback.

I believe this functionality will remain useful, I have no strong opinions on the way to provide it.

Comment (niteria) [reply jsbits/isomorphic.js:34]: If we disagree, can we move the discussion to #588 and proceed with the main contents of this pull request? This is a really nice fix and I wouldn't want to stall it on a tangential issue.

Comment (niteria) [reply jsbits/isomorphic.js:34]: Alternatively, we could rename `debug` to `logLevel` and have multiple logging levels: `0 - off`, `1 - error`, `2 - warn`, `3 - info`, `4 - debug`, `5 - trace`.

**Comment by dmjio** [comment]
@niteria @FPtje I've added some additional handling of `mountPoint`, that will ensure `miso` isn't mounted on a text node, and that miso will play nice alongside `<script>` that exist as siblings in `<body>`. I can rename `debug` to something more specific, i.e. `logLevel`. We should in theory be ready to go now.

**Comment by FPtje** [comment]
logLevel sounds good

**Comment by dmjio** [comment]
@niteria @ParthShah31 @tysonzero @BenjaminBaker @FPtje

I think this is ready to go (build is red due to `nix` issues), so disregard.

This has incorporated all of the aforementioned and is working well (tests pass and integration tested on some extremely large apps cc @ParthShah31). This really makes the isomorphic functionality first class, excited for things to become more user-friendly. Any last comments, concerns, cries of outrage before we merge?

**Comment by FPtje** [comment]
> cries of outrage

Well actually...

Just kidding 😂
LGTM! Nice work!

**Comment by niteria** [comment]
LGTM, but I expected `logLevel` to be a number, so that if someone only wants to see the errors they'd set it to `1` and the places that print errors would do `if (logLevel >= 1) ...` (in real code `1` hidden behind a constant, but shown here for demonstration). 

If someone wants to see the the debug messages they'd set it to `4` and the places with debug messages would do `if (logLevel >= 4) ...`. The nice thing is that the debug messages also enable error messages.

Might be overkill for now since we only have debug messages, but that's kind of what I had in mind.

**Comment by dmjio** [comment]
@niteria I agree with that approach, and I think we should adopt it as we add more logging facilities. For now, besides logging during prerendering, there isn't much else to log besides diffing. I didn't want to include extra JS log messages in diff.js because it would slow it down. Another use case for logging would be component mounting unmounting, but components right now are in Haskell land only, and have no js representation. So that's not a good use case for logging yet. But I agree, we can refactor add more when the need arises.

**Comment by tysonzero** [comment]
Would it be possible to have a separate log level that only shows warnings and not successes.

We won't really want to log "Successfully prendered page", particularly in production, but keeping the warnings around when things go wrong is definitely useful.

I know persistent uses [LogLevel](https://hackage.haskell.org/package/monad-logger-0.3.32/docs/Control-Monad-Logger.html#t:LogLevel) for this kind of thing. Presumably the success message would be around log level `info` or `debug`, and the warning message would be at log level `warn`.

**Comment by dmjio** [comment]
@tysonzero I agree there should be additional granularity when it comes to prerendering and logging. I probably should not have added logging in this PR to begin with, since this PR now conflates a few things, but it was strategic for my own debugging. I think we should open a new issue related to logging granularity to fulfill this request, and potentially logging events as well cc @niteria

**Pull request merged**